web_app/utils/data_source.py

```diff
@@ -0,0 +1,365 @@
+"""REST interface for reading and writing MoU data."""
+
+
+from copy import deepcopy
+from typing import cast, List, Optional, Tuple
+
+import pandas as pd  # type: ignore[import]
+
+from .icecube_setup import ICECUBE_INSTS
+from .types import Record, Table
+
+# Constants
+LABOR_CAT_LABEL = "Labor Cat."
+INSTITUTION_LABEL = "Institution"
+
+
+_ID = "id"
+_WBS_L2 = "WBS L2"
+_WBS_L3 = "WBS L3"
+_US_NON_US = "US / Non-US"
+_NAMES = "Names"
+_TASKS = "Tasks"
+_SOURCE_OF_FUNDS_US_ONLY = "Source of Funds (U.S. Only)"
+_FTE = "FTE"
+_NSF_MO_CORE = "NSF M&O Core"
+_NSF_BASE_GRANTS = "NSF Base Grants"
+_US_INSTITUTIONAL_IN_KIND = "U.S. Institutional In-Kind"
+_EUROPE_ASIA_PACIFIC_IN_KIND = "Europe & Asia Pacific In-Kind"
+_GRAND_TOTAL = "Grand Total"
+
+
+_US = "US"
+_NON_US = "Non-US"
+
+
+# read data from excel file
+_TABLE = pd.read_excel("WBS.xlsx").fillna("")
+_TABLE = [r for r in _TABLE.to_dict("records") if any(r.values())]
+_TABLE = {f"Z{r[_ID]}": r for r in _TABLE}
+for key in _TABLE.keys():
+    _TABLE[key][_ID] = key
+
+
+# --------------------------------------------------------------------------------------
+# Data/Table functions
+
+
+def pull_data_table(institution: str = "", labor: str = "") -> Table:
+    """Get table, optionally filtered by institution and/or labor."""
+    table = list(deepcopy(_TABLE).values())  # very important to deep-copy here
+
+    # filter by labor
+    if labor:
+        table = [r for r in table if r[LABOR_CAT_LABEL] == labor]
+
+    # filter by institution
+    if institution:
+        table = [r for r in table if r[INSTITUTION_LABEL] == institution]
+
+    def _us_or_non_us(institution: str) -> str:
+        for inst in ICECUBE_INSTS.values():
+            if inst["abbreviation"] == institution:
+                if inst["is_US"]:
+                    return _US
+                return _NON_US
+        return ""
+
+    # don't use US/Non-US from excel b/c later this won't even be stored in the DB
+    for record in table:
+        record[_US_NON_US] = _us_or_non_us(record[INSTITUTION_LABEL])
+
+    for record in table:
+        for field in record.keys():
+            if record[field] is None:
+                record[field] = ""
+
+    # sort
+    table = sorted(
+        table,
+        key=lambda k: (
+            k[_WBS_L2],
+            k[_WBS_L3],
+            k[_US_NON_US],
+            k[INSTITUTION_LABEL],
+            k[LABOR_CAT_LABEL],
+            k[_NAMES],
+            k[_SOURCE_OF_FUNDS_US_ONLY],
+        ),
+    )
```

Comment:
consider sorting in-place with `table.sort()`

Reply:
good catch